core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -168,39 +173,56 @@ public long getSizeInBytes()
         return sizeInBytes;
     }
 
-    private void copyBytes(Block block, int[] lengths, int[] positions, int count, int[] targetOffsets, int targetOffsetsIndex, int newByteCount)
+    private void copyBytes(Slice rawSlice, int[] lengths, int[] sourceOffsets, int count, int newByteCount)
     {
         ensureBytesCapacity(getCurrentOffset() + newByteCount);
 
-        for (int i = 0; i < count; i++) {
-            int position = positions[i];
-            if (!block.isNull(position)) {
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
+            for (int i = 0; i < count; i++) {
                 int length = lengths[i];
-                Slice slice = block.getSlice(position, 0, length);
-                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+                if (length > 0) {
+                    int offset = sourceOffsets[i];
+                    System.arraycopy(base, offset, bytes, offsets[positionCount + i], length);
+                }
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                int length = lengths[i];
+                if (length > 0) {
+                    int offset = sourceOffsets[i];
+                    rawSlice.getBytes(offset, bytes, offsets[positionCount + i], length);
+                }
             }
         }
-
         positionCount += count;
         updateSize(count, newByteCount);
     }
 
     /**
-     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     * Copy {@code length} bytes from {@code rawSlice}, starting at offset {@code sourceOffset} to {@code count} consecutive positions in the {@link #bytes} array.
      */
-    private void duplicateBytes(Block block, int position, int count)
+    private void duplicateBytes(Slice rawSlice, int sourceOffset, int length, int count)
     {
-        int length = block.getSliceLength(position);
         int newByteCount = toIntExact((long) count * length);
         int startOffset = getCurrentOffset();
         ensureBytesCapacity(startOffset + newByteCount);
 
-        Slice slice = block.getSlice(position, 0, length);
+        Object base = rawSlice.getBase();
+        if (base instanceof byte[] && rawSlice.getAddress() == ARRAY_BYTE_BASE_OFFSET) {
+            for (int i = 0; i < count; i++) {
+                System.arraycopy(base, sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                rawSlice.getBytes(sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
```

Comment:
An alternative way here is to copy only once into the byte array from Slice using this method, then use `System.arrayCopy` to copy bytes within the byte array with doubling size at each step, so you get only log_2(count) calls to `System.arrayCopy` overall.
Not sure if it'll be faster though.

Reply:
This is a good idea.
First of all, it allows us to not rely on Slice being backed by a byte array.
Second, I did a quick benchmark and this method is ridiculously fast on small to medium chunk sizes.
I suspect it's not gonna be that used in practice but I included this in this PR given results below.
```
Benchmark                               (length)  Mode  Cnt      Score   Error  Units
BenchmarkDuplicateBytes.arrayCopy              2  avgt   20      4.101 ±   0.082 ns/op
BenchmarkDuplicateBytes.arrayCopy             16  avgt   20      5.088 ±   0.055 ns/op
BenchmarkDuplicateBytes.arrayCopy             35  avgt   20      4.389 ±   0.175 ns/op
BenchmarkDuplicateBytes.arrayCopy             32  avgt   20      4.333 ±   0.049 ns/op
BenchmarkDuplicateBytes.arrayCopy             64  avgt   20      4.900 ±   0.050 ns/op
BenchmarkDuplicateBytes.arrayCopy             80  avgt   20      5.609 ±   0.117 ns/op
BenchmarkDuplicateBytes.arrayCopy            128  avgt   20      5.318 ±   0.069 ns/op
BenchmarkDuplicateBytes.arrayCopy            200  avgt   20      6.087 ±   0.098 ns/op
BenchmarkDuplicateBytes.arrayCopy            256  avgt   20      6.971 ±   0.111 ns/op
BenchmarkDuplicateBytes.arrayCopy            512  avgt   20     12.175 ±   0.324 ns/op
BenchmarkDuplicateBytes.arrayCopy           1024  avgt   20     22.759 ±   0.300 ns/op
BenchmarkDuplicateBytes.arrayCopy           2048  avgt   20     55.769 ±   1.213 ns/op
BenchmarkDuplicateBytes.arrayCopy         100000  avgt   20   4931.827 ±  276.57 ns/op
BenchmarkDuplicateBytes.arrayCopy        1000000  avgt   20  70111.621 ± 922.907 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x         2  avgt   20      0.230 ±   0.005 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        16  avgt   20      0.352 ±   0.015 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        35  avgt   20      0.533 ±   0.052 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        32  avgt   20      0.498 ±   0.007 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        64  avgt   20      0.881 ±   0.013 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x        80  avgt   20      1.100 ±   0.015 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       128  avgt   20      1.518 ±   0.228 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       200  avgt   20      3.522 ±   0.069 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       256  avgt   20      4.472 ±   0.154 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x       512  avgt   20     10.083 ±   0.206 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x      1024  avgt   20     21.741 ±   0.349 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x      2048  avgt   20     57.329 ±   0.854 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x    100000  avgt   20   5243.664 ± 188.963 ns/op
BenchmarkDuplicateBytes.copyFromSelf2x   1000000  avgt   20 109400.739 ± 992.325 ns/op
```